lib/RecommendationProvider.js

```diff
@@ -0,0 +1,157 @@
+/* globals exports, require, Task */
+"use strict";
+const {Cu} = require("chrome");
+const {setTimeout, clearTimeout} = require("sdk/timers");
+const simplePrefs = require("sdk/simple-prefs");
+
+const MILLISECONDS_IN_A_MINUTE = 1000 * 60;
+const MILLISECONDS_IN_TEN_MINUTES = MILLISECONDS_IN_A_MINUTE * 10;
+const MILLISECONDS_IN_AN_HOUR = MILLISECONDS_IN_A_MINUTE * 60;
+const MILLISECONDS_IN_A_DAY = MILLISECONDS_IN_AN_HOUR * 24;
+const MILLISECONDS_IN_A_WEEK = MILLISECONDS_IN_A_DAY * 7;
+const POCKET_PREF = "pocket.endpoint";
+
+const DEFAULT_TIMEOUTS = {
+  pocketTimeout: MILLISECONDS_IN_TEN_MINUTES, // every 10 minutes, refresh the Pocket recommendations
+};
+
+Cu.import("resource://gre/modules/Task.jsm");
+Cu.importGlobalProperties(["fetch"]);
+
+function RecommendationProvider(previewProvider, options = {}) {
+  this.options = Object.assign({}, DEFAULT_TIMEOUTS, options);
+  this._recommendedContent = [];
+  this._blockedRecommendedContent = [];
+  this._currentRecommendation = null;
+  this._pocketEndpoint = simplePrefs.prefs[POCKET_PREF];
+  this._previewProvider = previewProvider;
+  this._asyncExpireRecommendations(this.options.pocketTimeout);
+}
+
+RecommendationProvider.prototype = {
+  _pocketTimeoutID: null,
+
+  /**
+    * If we need a recommendation, go get one
+    */
+  getARecommendation(getRecommendation) {
+    if (getRecommendation) {
+      // if we are already displaying a recommendation, just show that one and don't get a new one
+      if (this._currentRecommendation) {
+        return this._currentRecommendation;
+      }
+      return this._getRandomRecommendation();
+    }
+    // if we don't need a recommendation, don't return one
+    return null;
+  },
+
+  /**
+    * Get a new recommendation from Pocket
+    */
+  getANewRecommendation(getRecommendation) {
+    this._currentRecommendation = null;
+    this.getARecommendation(getRecommendation);
+  },
+
+  /**
+    * Randomly pick a recommendation from the allowed set of recommendations (i.e ones that are not blocked)
+    */
+  _getRandomRecommendation() {
+    const allowedRecommendations = this._recommendedContent.filter(item => !this._blockedRecommendedContent.includes(item.url));
```

Comment:
using a `Set` would make this an amortized `O(1 * n)` operation, where n is the number of recommendations available. Using an `Array` makes it a `O(n * m)` operation where `m` is the number of blocked items and `n` is the number of recommendations.

Again, as stated above, the numbers `m` and `n` will not be big, but it's still more correct to use a `Set` here.